src/gui/sdlmain.cpp

```diff
@@ -1846,19 +1950,43 @@ static void SetupWindowResolution(const char *val)
 		SDL_Rect bounds;
 		SDL_GetDisplayBounds(sdl.display_number, &bounds);
 		if (w > 0 && h > 0 && w <= bounds.w && h <= bounds.h) {
+			sdl.desktop.want_resizable_window = false;
 			sdl.desktop.window.use_original_size = false;
 			sdl.desktop.window.width = w;
 			sdl.desktop.window.height = h;
 			return;
 		}
 	}
 
-	LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
-	        "'original' instead",
-	        pref.c_str());
+        LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
+                "'original' instead", pref.c_str());
+        sdl.desktop.want_resizable_window = false;
 	sdl.desktop.window.use_original_size = true;
 }
 
+static SDL_Rect CalculateViewport(int win_width, int win_height)
+{
+	const double prog_aspect_ratio = (sdl.draw.width * sdl.draw.scalex) /
+	                                 (sdl.draw.height * sdl.draw.scaley);
+	const double win_aspect_ratio = double(win_width) / double(win_height);
+
+	if (prog_aspect_ratio > win_aspect_ratio) {
+		// match window width
+		const int w = win_width;
+		const int h = win_width / prog_aspect_ratio;
+		const int x = 0;
+		const int y = (win_height - h) / 2;
+		return {x, y, w, h};
+	} else {
+		// match window height
+		const int w = win_height * prog_aspect_ratio;
+		const int h = win_height;
+		const int x = (win_width - w) / 2;
+		const int y = 0;
```

Comment:
Similar here; but `x` has less of a change of being negative (only happens if `win_height * prog_aspect_ratio` manages to exceed `win_width`).